engine/src/flutter/shell/platform/darwin/macos/framework/Source/ResizeSynchronizer.swift

```diff
@@ -0,0 +1,151 @@
+///
+/// When a native window containing a Flutter view is resized, the platform's window bounds can
+/// change before Flutter has been able to render a new frame matching those new dimensions. This
+/// asynchronicity can lead to undesirable visual effects such as:
+///   - Content temporarily appearing at the old size within the new, larger bounds.
+///   - Black "letterboxing" artifacts rendered at view edges.
+///   - A perception of lag in content resizing.
+///
+/// `ResizeSynchronizer` mitigates these issues by introducing a controlled blocking mechanism
+/// during the resize lifecycle:
+///
+/// 1.  Initiation (Platform Thread): When a native window resize is detected (e.g., by a window
+///     delegate), the platform thread calls `beginResize(forSize:notify:onTimeout:)` with the new
+///     target dimensions. This method blocks the calling platform thread, pausing the completion of
+///     the native resize event. While blocked, it continuously polls for engine messages to process
+///     relevant updates. The `notify` closure is invoked just before blocking begins.
+///
+/// 2.  Commit (Raster Thread): As the engine renders frames, the raster thread, upon presenting a
+///     new frame, calls `performCommit(forSize:afterDelay:notify:)`, providing the size of the
+///     frame just rendered.
+///
+/// 3.  Unblocking: If the size of the committed frame matches the target size that `beginResize()`
+///     is waiting for, `beginResize()` unblocks. This allows the native resize operation to
+///     complete, now that Flutter's content is synchronized with the new window dimensions.
+///
+/// Safeguards:
+///   - Timeout: `beginResize()` includes a timeout mechanism to prevent indefinite blocking if a
+///     matching frame isn't committed in a timely manner. An optional `onTimeout` closure can be
+///     provided to handle this event.
+///   - Shutdown: The synchronization can be cleanly interrupted by calling `shutDown()`, which will
+///     also unblock any pending `beginResize()` call. After shutdown, `beginResize()` will no
+///     longer block.
+///   - Thread Safety: The class manages its internal state in a thread-safe manner to coordinate
+///     actions between the platform thread and the raster thread.
+@objc(FlutterResizeSynchronizer)
+public final class ResizeSynchronizer: NSObject {
+  private static let invalidSize = CGSize(width: -1, height: -1)
+
+  // Synchronizes access to _isInResize_unsafe: isInResize is accessed from multiple threads and
+  // thus requires synchronized access to the underlying storage.
+  private let isInResizeLock = NSLock()
+  private var _isInResize_unsafe: Bool = false
+  private var isInResize: Bool {
+    get {
+      isInResizeLock.lock()
+      defer { isInResizeLock.unlock() }
+      return _isInResize_unsafe
+    }
+    set {
+      isInResizeLock.lock()
+      _isInResize_unsafe = newValue
+      isInResizeLock.unlock()
```

Comment:
Just curious, why not use `defer` here, like in the `get`er? I don't know Swift, but in other languages the `defer` approach would ensure we unlock if there's an exception.

Reply:
It's necessary in the getter since you can't unlock after a return statement otherwise. There's no exception that can occur in the assignment in the setter, so using defer maps to exactly the code here, so figured why not just write it out explicitly. 

I'd be fine with the other way too, but this felt more readable to me.

Reply:
This might be a @hellohuanlin question then on what's the idiomatic Swift pattern :)